File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B07_outliers_lof_isolationforest/R/B07_outliers_lof_isolationforest.R
# ############################################################################
# B7. OUTLIERS: LOF
# Tema B07_outliers_lof_isolationforest — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B07_outliers_lof_isolationforest.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B7. OUTLIERS: LOF\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B7. OUTLIERS: LOF E ISOLATION FOREST
# ----------------------------------------------------------------------------
# Estos dos atacan el problema desde ángulos OPUESTOS.
#
# LOF (Local Outlier Factor)
#   Compara la DENSIDAD de un punto con la de sus vecinos. Si vive en una
#   zona mucho más vacía que sus vecinos, es outlier.
#     LOF ~ 1  densidad normal
#     LOF >> 1 OUTLIER
#     LOF < 1  dentro de un cluster denso
#   Detecta outliers LOCALES: normales globalmente, raros en su vecindario.
#   ⚠️ En R (dbscan::lof) MAYOR = más atípico.
#      En Python (negative_outlier_factor_) es al revés.
#
# ISOLATION FOREST
#   Idea distinta: LOS PUNTOS RAROS SON FÁCILES DE AISLAR. Hace cortes
#   aleatorios; un outlier queda solo tras pocos cortes. Detecta outliers
#   GLOBALES. Es O(n log n) -> escala muy bien.
#   ⚠️ En R (isotree) MAYOR = más anómalo. En Python decision_function al revés.
#
# CUÁL USAR
#   n grande, muchas variables        -> ISOLATION FOREST
#   densidades muy distintas          -> LOF
#   pocas variables, datos ~normales  -> MAHALANOBIS (B6)
#   EN LA PRÁCTICA: corre varios y mira dónde COINCIDEN.
#
# PARÁMETROS
#   minPts (LOF)    : ⚠️ EL CRÍTICO. 10-35. Pequeño=ruidoso, grande=global
#   ntrees          : 100 basta
#   sample_size     : 256 funciona tan bien como usar todo
# ============================================================================

library(dbscan); library(isotree)

score_lof <- lof(Xs, minPts=20)                    # ⬅ MAYOR = más atípico
es_lof    <- score_lof > 1.5                       # umbral habitual (arbitrario)

iso      <- isolation.forest(as.data.frame(Xs), ntrees=100, sample_size=256,
                             nthreads=1, seed=0)
score_if <- predict(iso, as.data.frame(Xs))        # ⬅ MAYOR = más anómalo
es_if    <- score_if > quantile(score_if, 0.90)

cat(sprintf("outliers según LOF            : %d\n", sum(es_lof)))
cat(sprintf("outliers según Isolation Forest: %d\n", sum(es_if)))
cat(sprintf("\ncoinciden en             : %d\n", sum(es_lof & es_if)))
cat(sprintf("solo LOF (LOCALES)       : %d\n", sum(es_lof & !es_if)))
cat(sprintf("solo iForest (GLOBALES)  : %d\n", sum(!es_lof & es_if)))

par(mfrow=c(1,3))
plot(Z2, col=ifelse(es_lof,"red","gray60"), pch=16, xlab="PC1", ylab="PC2", main="LOF")
plot(Z2, col=ifelse(es_if,"red","gray60"),  pch=16, xlab="PC1", ylab="PC2", main="Isolation Forest")
plot(Z2, col="gray80", pch=16, xlab="PC1", ylab="PC2", main="Dónde coinciden")
points(Z2[es_lof & es_if, , drop=FALSE],  col="red",       pch=16, cex=1.4)
points(Z2[es_lof & !es_if, , drop=FALSE], col="blue",      pch=17, cex=1.4)
points(Z2[!es_lof & es_if, , drop=FALSE], col="darkgreen", pch=15, cex=1.4)
legend("topright", c("ambos","solo LOF","solo iForest"),
       col=c("red","blue","darkgreen"), pch=c(16,17,15), cex=.8)
par(mfrow=c(1,1))

cat("\nListo. Revisa la carpeta figuras/\n")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B07_outliers_lof_isolationforest/python/B07_outliers_lof_isolationforest.py
"""
B7. Outliers: LOF e Isolation Forest
Tema B07_outliers_lof_isolationforest — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B07_outliers_lof_isolationforest.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B7. Outliers: LOF e Isolation Forest")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B7. LOF e Isolation Forest ═══

# ── LOF ────────────────────────────────────────────────────────────────
lof = LocalOutlierFactor(n_neighbors=20, contamination="auto")  # ⬅ n_neighbors: el crítico
y_lof = lof.fit_predict(Xs)                      # -1 = outlier, 1 = normal
score_lof = lof.negative_outlier_factor_         # ⬅ es -LOF: más negativo = más atípico

# ── Isolation Forest ───────────────────────────────────────────────────
iso = IsolationForest(n_estimators=100, max_samples=256,
                      contamination="auto", random_state=0).fit(Xs)
y_if = iso.predict(Xs)                           # -1 = outlier
score_if = iso.decision_function(Xs)             # ⬅ más BAJO = más anómalo

es_lof, es_if = y_lof==-1, y_if==-1
print(f"outliers según LOF            : {int(es_lof.sum())}")
print(f"outliers según Isolation Forest: {int(es_if.sum())}")
print(f"\ncoinciden en                  : {int((es_lof & es_if).sum())}")
print(f"solo LOF (LOCALES)            : {int((es_lof & ~es_if).sum())}")
print(f"solo iForest (GLOBALES)       : {int((~es_lof & es_if).sum())}")

fig, ax = plt.subplots(1, 3, figsize=(16, 4.2))
ax[0].scatter(Z2[:,0], Z2[:,1], c=es_lof, cmap="coolwarm", alpha=.7); ax[0].set_title("LOF")
ax[1].scatter(Z2[:,0], Z2[:,1], c=es_if, cmap="coolwarm", alpha=.7); ax[1].set_title("Isolation Forest")
ax[2].scatter(Z2[~es_lof & ~es_if,0], Z2[~es_lof & ~es_if,1], c="lightgray", s=25, label="normales")
ax[2].scatter(Z2[es_lof & es_if,0],   Z2[es_lof & es_if,1],   c="red", s=65, label="ambos")
ax[2].scatter(Z2[es_lof & ~es_if,0],  Z2[es_lof & ~es_if,1],  c="blue", marker="^", s=65, label="solo LOF")
ax[2].scatter(Z2[~es_lof & es_if,0],  Z2[~es_lof & es_if,1],  c="green", marker="s", s=65, label="solo iForest")
ax[2].legend(fontsize=8); ax[2].set_title("Dónde coinciden y dónde no")
for a in ax: a.set_xlabel("PC1"); a.set_ylabel("PC2")
plt.tight_layout(); plt.show()

print(f"\nFilas más anómalas según iForest: {np.argsort(score_if)[:8]}")

print("\nListo. Revisa la carpeta figuras/")
